homework1/task04/sum_of_four/task04.py

```diff
@@ -0,0 +1,26 @@
+"""
+Classic task, a kind of walnut for you.
+
+Given four lists A, B, C, D of integer values,
+compute how many tuples (i, j, k, l) there are such that A[i] + B[j] + C[k] + D[l] is zero.
+We guarantee, that all A, B, C, D have same length of N where 0 <= N <= 1000.
+"""
+
+from typing import List
+
+
+def check_sum_of_four(a: List[int], b: List[int], c: List[int], d: List[int]) -> int:
+    """Return the amount of tuples where sum of elements is zero."""
+    ab = []
+    cd = []
+    counter = 0
+    for i in range(len(a)):
+        for j in range(len(c)):
+            ab.append(a[i] + b[j])
+            cd.append(c[i] + d[j])
+
+    for i in ab:
+        for j in cd:
+            if i == 0 - j:
+                counter += 1
```

Comment:
What is the complexity of this code?  Can it be better?

Reply:
O(n**2) because you need to run through n elements in cd sequence and also through n elements where n = len(ab) or len(cd). I tried to decrease complexity, but it was not successful. To make code little, I only can offer Counter(i+j for i in ab for j in cd)[0]. But it will not decrease complexity.

Reply:
Try to google it.

Reply:
I understand that here in **_ab_** we have n**2 combinations if **_a_** and _**b**_ lists. I mean the length of ab array here is n**2. It is too big. The same for cd array. So, i suppose, it would be more efficient to use dict to avoid non-unique values  of **_a_** and _**b**_ sums and count their amount. And we need namely default dict not to have KeyError when there is no element in dict.

Reply:
> And we need namely default dict not to have KeyError when there is no element in dict.

`defaultdict` just removes some boilerplate, you can use `dict` here.